### ISSUE TITLE
docker-commit and ADD/COPY truncate file and directory modification times

### ISSUE THREAD

**Issue by gg7** [docker/docker#17175]
**Steps to reproduce**:

COPY:

```bash
gg7@gg7:~$ mkdir docker-mtime-test
gg7@gg7:~$ cd docker-mtime-test

gg7@gg7:~/docker-mtime-test$ touch example-file
gg7@gg7:~/docker-mtime-test$ stat example-file
  File: ‘example-file’
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: fc01h/64513d    Inode: 7602338     Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1001/  gg7)   Gid: ( 1001/  gg7)
Access: 2015-10-19 14:13:28.441785563 +0100
Modify: 2015-10-19 14:13:28.441785563 +0100
Change: 2015-10-19 14:13:28.441785563 +0100
 Birth: -

gg7@gg7:~/docker-mtime-test$ cat > Dockerfile <<EOF
FROM ubuntu:14.04
COPY example-file /example-file
RUN stat /example-file
EOF

gg7@gg7:~/docker-mtime-test$ docker build --no-cache -t docker-mtime-test .
Sending build context to Docker daemon  2.56 kB
Step 0 : FROM ubuntu:14.04
 ---> fa81ed084842
Step 1 : COPY example-file /example-file
 ---> 8212ccaf4c14
Removing intermediate container cece2b393603
Step 2 : RUN stat /example-file
 ---> Running in 86bd29f2fa49
  File: '/example-file'
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 6ch/108d        Inode: 43          Links: 1
Access: (0664/-rw-rw-r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2015-10-19 13:15:27.415707557 +0000
Modify: 2015-10-19 13:13:28.000000000 +0000
Change: 2015-10-19 13:15:27.415707557 +0000
 Birth: -
 ---> cb0e1fdb58bd
Removing intermediate container 86bd29f2fa49
Successfully built cb0e1fdb58bd
```

**Actual results**: the modification time of `/example-file` is reported to be `2015-10-19 13:13:28.000000000 +0000`.

**Expected results**: the modification time of `/example-file` should be `2015-10-19 13:13:28.441785563 +0000`.

`docker-commit`:
```bash
gg7@gg7:~$  docker run -it --name mtime-test ubuntu:14.04 /bin/bash
root@613e6c0fe6ac:/# touch /example-file
root@613e6c0fe6ac:/# stat /example-file
  File: '/example-file'
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 66h/102d        Inode: 72          Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2015-10-19 13:36:29.888088708 +0000
Modify: 2015-10-19 13:36:29.888088708 +0000
Change: 2015-10-19 13:36:29.888088708 +0000
 Birth: -
root@613e6c0fe6ac:/# exit
exit
gg7@gg7:~$ docker commit mtime-test mtime-test-image
b67d16fcb577fb3ece29cf007f1cef2ebc4a6351bc3e3a3e77eab5648461095c
gg7@gg7:~$ docker run -it --name mtime-test-2 mtime-test-image /bin/bash
root@1a5e038e6a2c:/# stat /example-file
  File: '/example-file'
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 66h/102d        Inode: 75          Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2015-10-19 13:36:55.572503813 +0000
Modify: 2015-10-19 13:36:29.000000000 +0000
Change: 2015-10-19 13:36:55.572503813 +0000
 Birth: -
root@1a5e038e6a2c:/#
```

**Why is this important?**

Some build systems rely on modification timestamps to detect file updates.

**Environment details**:

`cat /etc/lsb-release`:
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.3 LTS"
```

`uname -a`:
```
Linux gg7 3.13.0-51-generic #84-Ubuntu SMP Wed Apr 15 12:08:34 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

`docker version`:
```
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 05:37:18 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 05:37:18 UTC 2015
 OS/Arch:      linux/amd64
```

`docker info`:
```
Containers: 8
Images: 632
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 656
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-51-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 8
Total Memory: 15.66 GiB
Name: gg7
ID: ACSH:EYI5:G36O:2SRA:FWLH:U4GF:VMJM:C4KF:73MH:L273:TCCO:GW3K
WARNING: No swap limit support
```

**Comment by schnittchen** [comment]
Just ran into this today.

The only conceivable workaround is taring files up, COPYing them and extracting again...

**Comment by zyxue** [comment]
Given this issue has been open for so long, I assume it can be hard to fix in Docker. I'm curious why it would be hard to preserve the timestamp during `COPY`?

**Comment by daym** [comment]
This bug makes it basically impossible to select only parts you need from an existing stage using "COPY --from"--because every time you do, it messes something up in the subset.

Since usually if you compose things you do NOT want to have tools like tar and co available (except when you are the author of tar, I guess), the tar workaround is a hack, too.

**Comment by mercuriete** [comment]
I am hitting this problem in this scenario:

I have the following infrastructure:

user -> Cloudfront -> EKS -> nextjs server.

When a user request an assets the user request the asset with:
```if-modified-since``` HTTP header.

If the expiration time on cloudfront is expired, cloudfront send to the backend the same cache HTTP headers to know if it could be a RefreshHit instead of a full failure.

What is the problem here?

after a deployment of a new container... all modification have changed but in reality the git modification time is very different and that asset have not changed at all. Leading to a failure in cache when it should be a RefreshHit.

I can investigate why chrome is not sending eTag but if my browser is not using eTag and is using ```ìf-modified-since``` is a problem that I can't fix.

TLDR;
modification time could improve cache hits and networking expenses.

**Comment by leorochael** [comment]
This issue is interacting badly with workflows where the context contains artifacts that were partially built before the Dockerfile runs, but which a tool like `make` checks the file timestamps to see if they need to be rebuilt.

Inside the Dockerfile build process, `make` will then wrongly come to the conclusion that some artifacts are older than their origin files and will try to rebuild them.

**Comment by hunjixin** [comment]
> This issue is interacting badly with workflows where the context contains artifacts that were partially built before the Dockerfile runs, but which a tool like `make` checks the file timestamps to see if they need to be rebuilt.
> 
> Inside the Dockerfile build process, `make` will then wrongly come to the conclusion that some artifacts are older than their origin files and will try to rebuild them.

have the met  same issue, only.

**Comment by thaJeztah** [comment]
related;

- https://github.com/moby/moby/issues/48447

**Comment by cen1** [comment]
Not entirely sure this question is 100% related to this issue but I couldn't find anything closer. Does COPY preserve file timestamps? Such an option would help achieve bit identical reproducibility goals.
